### PR TITLE
[InstCombine] Simplify max(min(abs()))

### PR DESCRIPTION
**Pull request by sergachev**
Remove smax from smax(smin(abs(x), y), 0) if y >= 0.

### CODE REVIEW

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
It is incorrect when `x==INT_MIN`: https://alive2.llvm.org/ce/z/X4TZPY

BTW, isn't it already handled by CVP/SCCP?

**Comment by goldsteinn** [comment]
> It is incorrect when `x==INT_MIN`: https://alive2.llvm.org/ce/z/X4TZPY

It is valid if you make INT_MIN ub.

**Comment by sergachev** [comment]
> BTW, isn't it already handled by CVP/SCCP?

It isn't handled at the moment: https://alive2.llvm.org/ce/z/5o8Y-m

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 1]
```diff
@@ -1770,6 +1770,14 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
       }
     }
 
+    // smax(smin(abs(x), y), 0) -> smin(abs(x), y) where y >= 0
+    if (IID == Intrinsic::smax &&
+        match(I0, m_SMin(m_c_Intrinsic<Intrinsic::abs>(m_Value(), m_Value()),
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:1775]: Still need to be `m_c_Intrinsic<..::abs>(m_Value(), m_One())` (And your tests will need toupdated correspondingly.)

**Comment by nikic** [comment]
> > BTW, isn't it already handled by CVP/SCCP?
> 
> It isn't handled at the moment: https://alive2.llvm.org/ce/z/5o8Y-m

It's not handled because the transform is incorrect. The variant with int_min_is_poison set *is* already handled: https://alive2.llvm.org/ce/z/Ah_8wT

**Pull request closed without merging**